### ISSUE TITLE
unknown type name intptr_t when using clang-cl

### ISSUE THREAD

**Issue by MoAlyousef** [fltk/fltk#752]
**Describe the bug**
Windows now provides a clang toolchain with clang-cl as a drop-in replacement for msvc. However some issues arise from the libcxx headers that are bundled with the clang toolchain. Trying to build FLTK with clang-cl raises 2 issues in FL/platform_types.H:
```
FL/platform_types.h(105,9): error: unknown type name 'intptr_t'
  typedef intptr_t fl_intptr_t;
          ^
FL/platform_types.h(106,9): error: unknown type name 'uintptr_t'
  typedef uintptr_t fl_uintptr_t;
```

The platform_types.H file has an ifdef _MSC_VER include stddef.h, else stdint.h.
I think we can safely add stdint.h as well for MSVC, since clang-cl actually defines _MSC_VER, and the extra include would be harmless on windows.

**To Reproduce**
- git clone FLTK.
- Generate a visual studio project file with CMake. It needs to be something recent:
```
cmake -Bbin -G"Visual Studio 17 2022" -T ClangCL
```
- build via cmake --build bin or via visual studio

**Expected behavior**
FLTK should build with clang-cl.

**Screenshots**
N/A

**FLTK Version**
 - Version: 1.4.0
 - Downloaded and self-built from: Git
 - If from Git, branch: master

**FLTK Configure / Build Options**
See above

**Operating System / Platform:**
Please be as precise as possible, e.g. "Linux: Ubuntu 20.04"
 - OS: Windows
 - OS Version: Windows 11

**Comment by erco77** [comment]
Ya the ifdef that prevents stdint.h for _MSC_VER might be for very old compilers. Perhaps it's important to identify which visual studio compiler introduced stdint.h and we should more tightly constrict that ifdef to those older versions.

**Comment by DBLouis** [comment]
I'm interested for this to be solved. Here is another source claiming "stdint.h is not shipped with [..] Visual Studio C++ products prior to Visual Studio 2010"

https://en.wikibooks.org/wiki/C_Programming/stdint.h#External_links

**Comment by Albrecht-S** [comment]
The best solution might be to add a CMake check, i.e. try to find `stdint.h` and include it only if found.

As a simple compromise we could change the _MSC_VER check to be true only for versions < VS 2010 as described above (thanks, @erco77 and @DBLouis). I propose the following patch:
```
diff --git a/FL/platform_types.h b/FL/platform_types.h
index 421a0b7c6..7b5e95c62 100644
--- a/FL/platform_types.h
+++ b/FL/platform_types.h
@@ -96,8 +96,8 @@ typedef opaque Fl_Timestamp;
 
 #ifdef _WIN64
 
-#if defined(_MSC_VER)
-# include <stddef.h>  /* M$VC */
+#if defined(_MSC_VER) && (_MSC_VER < 1600)
+# include <stddef.h>  /* stdint.h not available before VS 2010 (1600) */
 #else
 # include <stdint.h>
 #endif
```

This patch would only affect Visual Studio before VS-2010 which should be rarely used today.

@MoAlyousef Would this fix the issue for you?

**Comment by Albrecht-S** [comment]
> This patch would only affect Visual Studio before VS-2010 which should be rarely used today.

This statement was not correct. The change introduced by the patch would affect **_all_** Visual Studio versions **_since_** `VS-2010` and all future versions because these versions would now include `stdint.h` rather than `stddef.h` which is intended.

Visual Studio versions before `VS-2010` would not be affected and work as before.

**Comment by Albrecht-S** [comment]
I also checked the feasibility of adding a header check to CMake. This would require to define a macro in `FL/fl_config.h`, the **public** configuration header.

I'm hesitating to do this because we currently include only macros that are configuration specific (i.e. depend on chosen build options and not on system inspection). Hence I'd prefer the patch posted above.

**Comment by MoAlyousef** [comment]
I think checking the _MSC_VER in platform_types is a reasonable solution.

**Comment by Albrecht-S** [comment]
@MoAlyousef Thanks, can you please test the given patch for us, i.e. does this patch work with clang?

I don't have VS 2022 installed and can't test it right now. I wouldn't like to commit a potential solution that has not been tested. TIA.

**Comment by MoAlyousef** [comment]
Thanks @Albrecht-S 
I applied the patch and confirm it works.

**Comment by Albrecht-S** [comment]
OK, then I'd like to commit my patch unless someone finds a reason not to do it within the next 32 hours or so [1] ... and posts it here.

[1] That would be about Jul 20, 10:00 CEST (08:00 UTC).

**Comment by Albrecht-S** [comment]
Fixed in commit 069a6b430b248aace1ec0f1a72099e95f046a68d.

Thanks for the report and for all contributions.
